File: 0216-combination-sum-iii/0216-combination-sum-iii.java
class Solution {
    static void findCombinations(int idx,int k,int n,List<List<Integer>> result,List<Integer> combination){
        if(k == 0 ){
            if(n == 0){
                result.add(new ArrayList<>(combination));
            }
            return;
        }

        for(int i=idx; i<10; i++){
            if(i <= n){
                combination.add(i);
                findCombinations(i+1,k-1,n-i,result,combination);
                combination.remove(combination.size()-1);
            }
        }
    
    }
    public List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>> result = new ArrayList<>();
        findCombinations(1,k,n,result,new ArrayList<>());
        return result;
    }
}

File: 0216-combination-sum-iii/0216-combination-sum-iii.js
/**
 * @param {number} k
 * @param {number} n
 * @return {number[][]}
 */
var combinationSum3 = function(k, n) {
    const result = [];

    backtrack(1, k, n, [], result);

    return result;
};

function backtrack(start, k, target, current, result) {

    // Found a valid combination
    if (target === 0 && current.length === k) {
        result.push([...current]);
        return;
    }

    // Invalid state
    if (target < 0 || current.length > k) {
        return;
    }

    // Try all numbers from 'start' to 9
    for (let i = start; i <= 9; i++) {
        current.push(i);

        // Each number can be used only once
        backtrack(i + 1, k, target - i, current, result);

        // Backtrack
        current.pop();
    }
}